Create Workflows to Upload new Releases

danyeaw opened the pull request:
Initial PR for #281.

This PR creates two new GitHub Actions workflows:

1. Drafts a release when a new version tag is pushed.
2. Uploads a release to PyPI once the release is published.

In order to take advantage of this, PyPI credentials need to be stored as [secrets](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets). The preferred way to do this would be to create a [PyPI API token](https://pypi.org/help/#apitoken), but your username and password could also be used.

freakboy3742 commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,26 @@
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install setuptools wheel twine
```
These probably need to be `--upgrade` too; setuptools *should* be installed in a default environment.

freakboy3742 commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,26 @@
+    - name: Set up Python
+      uses: actions/setup-python@v1
+      with:
+        python-version: '3.x'
```
TIL 3.x was possible in Github Actions...

freakboy3742 commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,24 @@
+          tag_name: ${{ github.ref }}
+          release_name: Release ${{ github.ref }}
+          draft: true
+          prerelease: false
```
Ideally, we'd set this dependent on whether the version tag has 'a', 'b', 'rc', 'dev', or 'pre' in it... is that possible? Or do we just have to keep our wits about us and make sure we uncheck this if we tag a dev release?

freakboy3742 commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,24 @@
+on:
+  push:
+    tags:
+      - 'v*'
```
It's an unlikely edge case, but if we tagged "very_important_milestone", this would trigger this workflow.  If I'm reading the docs right, a pattern of `'v[0-9]+.[0-9]+*'` would give us at least some superficial protection against this.

danyeaw replied:
I think the risk here and for the next comment is low. If an extra tag was pushed, it would just create a draft release. We would just need to go to the release tab and click on the delete draft button to clean it up. All the tags in the repo right now are for releases. I could add in the regular expression, but might be overkill to prevent an extra draft release for some one off event.

freakboy3742 replied:
Fair enough. As long as it's easy to clean up drafts that get mistakenly created, I agree that the extra complexity probably isn't worth it.

danyeaw replied:
Yup, it is very easy to clean them up, a single button click. :+1:

danyeaw commented:
@freakboy3742 Thanks for the feedback.

> We might want to set this configuration up on a test repository (and publish to testPyPI) to validate the processes will work as expected before we try to land this into a live project.

I could test this on my fork before we merge. Do you think I could get test-pypi write access?

I add the `twine check` command :+1:.
 
> Lastly - what happens if a tag is updated?

By updated, I think you mean for example the v1.0 tag is deleted and a new v1.0 tag is created and pushed? A new draft release would be created, and we would just have to delete the previous draft release on the Releases tab.

freakboy3742 commented:
Yeah - that's what I meant about deleting or changing a version tag.

I've just given you maintainer access on test.PyPI.org for briefcase.

Assuming the `twine check` fails... what's the recovery procedure? We can't retag the same hash, and the release will have already been published. Would it be better to do the `twine check` as a step on the tag event? That way, at least if it fails, we'll have to create an updated tag anyway. 

Alternatively, we could add `twine check` as a normal build check (i.e., as another pycodestyle check). That way, we hopefully won't ever end up accepting merging a PR that has a potential publication problem baked into the markup.

danyeaw commented:
@freakboy3742 I have completed my testing, and found one issue with the `twine check` command that I have fixed.

You can see the successful result here:
https://test.pypi.org/project/briefcase/0.3.0.dev2/
The upload is the result of my PR:
https://github.com/danyeaw/briefcase/pull/1
and this release:
https://github.com/danyeaw/briefcase/releases/tag/v0.3.0.dev2
I hope it is OK that I burned the 0.3.0.dev2 version on the test server.

I also added an error to the README and tried to publish a 0.3.0.dev3 version. As soon as the `twine check` command failed, the workflow immediately cancelled and the version was not uploaded to test.pypi.org. You can then delete the tag (locally and on the repo) and this sets the Release back to draft automatically, then push a new tag to fix the issue.

I also made use of the PyPI API Token, and that worked well with using `__token__` as the username.

danyeaw commented:
Yup, that is a good summary of the process 👍

freakboy3742 commented:
Righto then - Lets 🚢!
